dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java

```diff
@@ -0,0 +1,424 @@
+import org.hisp.dhis.cache.CacheProvider;
+import org.hisp.dhis.category.CategoryCombo;
+import org.hisp.dhis.common.IdScheme;
+import org.hisp.dhis.common.IdSchemes;
+import org.hisp.dhis.dataelement.DataElement;
+import org.hisp.dhis.dxf2.common.ImportOptions;
+import org.hisp.dhis.dxf2.events.event.Event;
+import org.hisp.dhis.organisationunit.FeatureType;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramStage;
+import org.hisp.dhis.program.ProgramStageDataElement;
+import org.hisp.dhis.program.ProgramType;
+import org.hisp.dhis.program.ValidationStrategy;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.user.sharing.Sharing;
+import org.springframework.core.env.Environment;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Component;
+
+/**
+ * This supplier builds and caches a Map of all the Programs in the system.
+ * For each Program, the following additional data is retrieved:
+ *
+ *
+ *
+ * @formatter:off
+ *
+         Program
+         +
+         |
+         +---+ Program Stage (List)
+         |           |
+         |           +---+ User Access (ACL)
+         |           |
+         |           +---+ User Group Access (ACL)
+         |
+         |
+         +---+ Category Combo
+         |
+         |
+         +---+ Tracked Entity Instance
+         |           |
+         |           +---+ User Access (ACL)
+         |           |
+         |           +---+ User Group Access (ACL)
+         |
+         |
+         |
+         +---+ Organizational Unit (List)
+         |
+         +---+ User Access (ACL)
+         |
+         +---+ User Group Access (ACL)
+ * @formatter:on
+ *
+ * @author Luciano Fiandesio
+ */
+@Component( "workContextProgramsEventSupplier" )
+public class ProgramEventSupplier extends AbstractSupplier
+{
+    private static final String PROGRAM_CACHE_KEY = "000PES";
+
+    private static final String ATTRIBUTESCHEME_COL = "attributevalues";
+
+    private static final String PROGRAM_STAGE_ID = "programstageid";
+
+    private static final String COMPULSORY = "compulsory";
+
+    private static final String TRACKED_ENTITY_TYPE_ID = "trackedentitytypeid";
+
+    // Caches the entire program hierarchy, including program stages and ACL
+    private final Cache<Map<String, Program>> programsCache;
+
+    public ProgramEventSupplier( NamedParameterJdbcTemplate jdbcTemplate, CacheProvider cacheProvider,
+        Environment environment )
+    {
+        super( jdbcTemplate, environment );
+        programsCache = cacheProvider.createProgramCache();
+    }
+
+    public Map<String, Program> get( ImportOptions importOptions, List<Event> eventList )
+    {
+        boolean requiresReload = false;
+
+        // Do not use cache if {@code skipCache} is true or if running as test
+
+        if ( importOptions.isSkipCache() )
+        {
+            programsCache.invalidateAll();
```

Comment:
Does it make sense to invalidate cache if this import is not using the cache?
What if there are 2 concurrent imports? Will this clear the cache used by the other import?